### PR TITLE
Adam track texture allocations

### PR DESCRIPTION
**Pull request by adam-mapbox**
I'm going to write up a doc about this, but this is the start of some code to track our GL usage.  If you want to turn it on, go to gl.hpp and uncomment GL_TRACK.

### CODE REVIEW

**Comment by incanus** [comment]
Nice. @lucaswoj is this the incarnation that you used? You have a better sense of the usefulness of this PR. 

Also, I'm curious why things like `GEOJSONVT_VERSION` and `src/mbgl/map/annotation.hpp` are updated here — I'm thinking this is out of sync with current `master`. @adam-mapbox was the merge at https://github.com/mapbox/mapbox-gl-native/commit/9b99e26830d8aab50266304e596779e3f390302f maybe from a stale `master`? Try merging `origin/master` after a `git fetch` to ensure you've got the latest upstream. 

After we're happy with the look here, we should squash it down a bit to cut some of the commit noise. 

/cc @kkaefer for some :eyes: as well.

**Comment by lucaswoj** [comment]
Yes, this is based on a stale master or a fork of master and should be rebased.

@adam-mapbox and I used this in conjunction with a few other tools like `Instruments.app` and the shadowy [`Graphics Tools.app`](https://developer.apple.com/library/mac/technotes/tn2178/_index.html) to track down a couple GPU memory leaks. 

I'll let @adam-mapbox make the case for whenever or not this should be merged into `master` -- I'm not an expert in that area.

**Comment by adam-mapbox** [comment]
I do think it can/should be merged in.  As to the stale fork, I think I made a mistake and I merged instead of rebasing.  Anyone want to help me extricate myself from the current situation?

**Comment by lucaswoj** [comment]
Rebased and squashed for you @adam-mapbox. Previous SHA was 69765c5, if you need to back out my changes.

**Comment by incanus** [comment]
@lucaswoj Help him walk through it next time :wink: 

@adam-mapbox Looking better, but there are a couple things I think we should address. I agree that this should live in `master` when it's ready, though. 

- [ ] We should back out the accidental app token commit in `ios/app/app-info.plist`. Reason is we want developers to have an account and get in the flow of using their own account's token. 

- [ ] Unless I'm missing something, instead of `std::cout`, we should be using our own `mbgl::Log` facilities. First, for consistency, but second, we should be able to control this logging, which otherwise would occur in all uses of `Headless`, right? Even without `GL_TRACK`?

**Comment by adam-mapbox** [comment]
@incanus The reason I used std::cout instead of Log was that, when we were running it for @lucaswoj , through node, the Log messages weren't coming out.  We couldn't figure that out, so we went back to using cout.

**Comment by incanus** [comment]
Ah! Good call. Through node, that does seem likely. @jfirebaugh @mikemorris does that seem ok to you?

**Comment by jfirebaugh** [comment]
The `cout`s look to me like temporary debugging code that should be removed before this is merged.

**Comment by incanus** [comment]
@adam-mapbox Looking like some of the out-of-scope changes are back as of the merge in https://github.com/mapbox/mapbox-gl-native/commit/5bf16d9aca96338e63778f0a6acc0282b1148421. I can tell this by looking at [the files tab for this PR](https://github.com/mapbox/mapbox-gl-native/pull/2437/files) and seeing the overall changes that are proposed. This updates with each commit. 

You can also double-check the overall impact with something like `git diff origin/master..adam_track_texture_allocations`, which will show the differences introduced in this branch.

**Comment by incanus** [comment]
I can't seem to resolve the state of this branch anymore. @lucaswoj can you re-push your rebase from before? Then @adam-mapbox you will need to `git fetch; git reset --hard origin/adam_track_texture_allocations`, then reapply your commit of https://github.com/mapbox/mapbox-gl-native/commit/41edeafb10b80974a2e61f06f9c12f1b759fd276 before pushing again. You should not push any merge commits such as https://github.com/mapbox/mapbox-gl-native/commit/5bf16d9aca96338e63778f0a6acc0282b1148421.

**Comment by lucaswoj** [comment]
Pushing now -- overwriting 41edeafb10b80974a2e61f06f9c12f1b759fd276 with 0797402a174acfdf0fb61a3475ef4aa4ba5123ed.

**Comment by lucaswoj** [comment]
I'm not 100% on this, but I seem to remember us thinking `mbgl::Log` wasn't working when really our logging code wasn't being triggered at all.

**Comment by mikemorris** [comment]
@lucaswoj in the Node bindings, mbgl::Log streams to an EventEmitter on either the module or map (forget if it changed with Environments), you have to listen on that for events and then log output somewhere.

**Comment by incanus** [comment]
That explains the logging not working for @adam-mapbox and @lucaswoj, but it sounds like per @jfirebaugh that we should remove the logging entirely now.

**Comment by jfirebaugh** [comment]
We should remove the debug logging in `HeadlessView`. The logging in `gl.cpp` is ok.

File: ios/app/app-info.plist [hunk 1]
```diff
@@ -2,6 +2,8 @@
 <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
 <plist version="1.0">
 <dict>
+	<key>MGLMapboxAccessToken</key>
```
Comment (jfirebaugh) [inline ios/app/app-info.plist]: Revert this change, per https://github.com/mapbox/mapbox-gl-native/pull/2437#issuecomment-143886759.

File: src/mbgl/map/map.cpp [hunk 2]
```diff
@@ -11,6 +11,8 @@
 #include <mbgl/util/thread.hpp>
 #include <mbgl/util/math.hpp>
 
+#include <iostream>
```
Comment (jfirebaugh) [inline src/mbgl/map/map.cpp]: Not used.

File: src/mbgl/platform/gl.cpp [hunk 3]
```diff
@@ -55,6 +59,143 @@ void checkError(const char *cmd, const char *file, int line) {
...
+#undef glBindTexture
+#undef glDeleteTextures
+#undef glTexImage2D
+#undef glClear
+#undef glShaderSource
+#undef glBufferData
+#undef glBindBuffer
+#undef glDeleteBuffers
+#undef glBufferData
+static unsigned int current_used_bytes = 0;
```
Comment (jfirebaugh) [inline src/mbgl/platform/gl.cpp]: Please use camel case variables here and elsewhere in the change (`currentUsedBytes`).

File: src/mbgl/platform/gl.cpp [hunk 4]
```diff
@@ -55,6 +59,143 @@ void checkError(const char *cmd, const char *file, int line) {
...
+static std::map<GLint, GLsizeiptr> buffer_binding_to_size_map;
+static unsigned int current_used_buffer_bytes = 0;
+static unsigned int largest_amount_used_so_far = 0;
+
+static std::mutex g_debug_mutex;
 
+void mbx_glBindBuffer(GLenum target,
+                      GLuint buffer) {
+    g_debug_mutex.lock();
+    if(target == GL_ARRAY_BUFFER) {
```
Comment (jfirebaugh) [inline src/mbgl/platform/gl.cpp]: Space between `if` and `(` (and `for` and `(`, etc.)

File: src/mbgl/platform/gl.cpp [hunk 5]
```diff
@@ -55,6 +59,143 @@ void checkError(const char *cmd, const char *file, int line) {
...
+static GLuint current_element_array_buffer = 0;
+static std::map<GLint, GLsizeiptr> buffer_binding_to_size_map;
+static unsigned int current_used_buffer_bytes = 0;
+static unsigned int largest_amount_used_so_far = 0;
+
+static std::mutex g_debug_mutex;
 
+void mbx_glBindBuffer(GLenum target,
+                      GLuint buffer) {
+    g_debug_mutex.lock();
```
Comment (jfirebaugh) [inline src/mbgl/platform/gl.cpp]: Use `std::lock_guard` rather than manual `lock` / `unlock`.

**Comment by adam-mapbox** [comment]
The couts are *not* temporary debugging code.  They are sort of the whole point of this.  Please don't remove them.

**Comment by jfirebaugh** [comment]
@adam-mapbox I'm talking about the logging in HeadlessView like:

```
std::cout << "HeadlessView::HeadlessView" << std::endl;
```

We can't have this sort of debug logging in a production build: it would clutter the output in places like our headless rendering tests and server side GL rendering.

I agree that the logs in gl.cpp should remain.

**Comment by adam-mapbox** [comment]
Yes, right, of course.  That was an error.

**Comment by incanus** [comment]
test comment

**Comment by incanus** [comment]
Force-pushed to c958ee45abe978274c1c86b8517b2c86bc303c8f — forget to force push before fast-forward marging into `master`.

**Comment by incanus** [comment]
(Worked through rebasing etc. with @adam-mapbox on this branch locally).
